test/unit/account_test.rb

```diff
@@ -736,6 +612,206 @@ def test_smart_destroy_master
+
+      provider.destroy!
+
+      assert_equal 0, provider.provided_cinstances.reload.count
+      assert_equal 0, provider.provided_plans.reload.count
+      assert_equal 0, provider.bought_plans.reload.count
+    end
+
+    test "#smart_destroy a buyer" do
+      buyer = FactoryBot.create(:buyer_account, provider_account: provider)
+
+      buyer.smart_destroy
+      assert_raise(ActiveRecord::RecordNotFound) { buyer.reload }
+    end
+
+    test "#smart_destroy a provider" do
+      provider.smart_destroy
+      assert provider.reload.scheduled_for_deletion?
+    end
+
+    test "#smart_destroy master" do
+      master_account.smart_destroy
+      assert master_account.reload
+      assert_not master_account.reload.scheduled_for_deletion?
+    end
+
+    test "email provider when buyer is deleted on failure unstoring credit card" do
+      buyer = FactoryBot.create(:buyer_account, provider_account: provider)
+      FactoryBot.create(:payment_detail, account: buyer, buyer_reference: "buyer-2")
+      preferences = provider.first_admin.notification_preferences
+      preferences.enabled_notifications = ["credit_card_unstore_failed"]
+      preferences.user.save!
+
+      ::Sidekiq::Testing.inline! do
+        buyer.destroy!
+      end
+
+      assert_equal 1, ActionMailer::Base.deliveries.size
+      assert_includes ActionMailer::Base.deliveries.first.header["X-SMTPAPI"].value, "CreditCardUnstoreFailedEvent"
+      assert_raise(ActiveRecord::RecordNotFound) { buyer.reload }
+    end
+
+    test "email provider when buyer is deleted on error unstoring credit card" do
+      buyer = FactoryBot.create(:buyer_account, provider_account: provider)
+      FactoryBot.create(:payment_detail, account: buyer, buyer_reference: "buyer-3")
+      preferences = provider.first_admin.notification_preferences
+      preferences.enabled_notifications = ["credit_card_unstore_failed"]
+      preferences.user.save!
+
+      ::Sidekiq::Testing.inline! do
+        buyer.destroy!
+      end
+
+      assert_equal 1, ActionMailer::Base.deliveries.size
+      assert_includes ActionMailer::Base.deliveries.first.header["X-SMTPAPI"].value, "CreditCardUnstoreFailedEvent"
+      assert_raise(ActiveRecord::RecordNotFound) { buyer.reload }
+    end
+
+    test "delete a provider with buyer credit cards" do
+      buyers = FactoryBot.create_list(:buyer_account, 3, provider_account: provider)
+      buyers.each_with_index do |buyer, idx|
+        FactoryBot.create(:payment_detail, account: buyer, buyer_reference: "buyer-#{idx}")
+      end
+      provider.schedule_for_deletion!
+
+      # cards of buyers of deleted providers should not be touched
+      ActiveMerchant::Billing::BogusGateway.any_instance.expects(:unstore).never
+
+      ::Sidekiq::Testing.inline! do
+        provider.destroy!
+      end
+
+      # TODO: master should not be emailed about events related to schduled_for_deletion providers
+      # assert_empty ActionMailer::Base.deliveries
```

Comment:
these TODO items are for another PR as I didn't want to mess with master notifications in this one